### PR TITLE
Add support for ADDRESS

### PR DESCRIPTION
**Pull request by thilgen**
### Context
`ADDRESS` is a formula provided by 
* **Google Sheets** ([documentation](https://support.google.com/docs/answer/3093308))
* **MS Excel** ([documentation](https://support.microsoft.com/en-us/office/address-function-d0c26c0d-3991-446b-8de4-ab46431d4f89))
* **Open Document Format for Office Applications** ([documentation](http://docs.oasis-open.org/office/v1.2/os/OpenDocument-v1.2-os-part2.html#__RefHeading__1018416_715980110))

### For more information
* Issue Writeup: https://github.com/handsontable/hyperformula/issues/1221

### How has this been tested?
Unit Tests (100% line coverage for new files and new paths)
```
error-message.ts                          |     100 |      100 |     100 |     100 |                                                         
AddressPlugin.ts                          |     100 |     90.9 |     100 |     100 | 47                                                      
index.ts                                  |     100 |      100 |     100 |     100 |                                                         
```
_(update 2023-04-28)_
```
error-message.ts                          |   98.64 |      100 |      75 |     100 |                                                                                                                                                                          
AddressPlugin.ts                          |     100 |      100 |     100 |     100 |                                                                                                                       
index.ts                                  |     100 |      100 |     100 |     100 |                                                         
```

### Types of changes

- [ ] Breaking change (a fix or a feature because of which an existing functionality doesn't work as expected anymore)
- [X] New feature or improvement (a non-breaking change that adds functionality)
- [ ] Bug fix (a non-breaking change that fixes an issue)
- [ ] Additional language file, or a change to an existing language file (translations)
- [ ] Change to the documentation

### Related issues:
1. https://github.com/handsontable/hyperformula/issues/1221

### Checklist:

- [ ] My change is compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard.
- [X] My change is compatible with Microsoft Excel.
- [X] My change is compatible with Google Sheets.
- [X] My code follows the code style of this project.
- [ ] I described my changes in the [CHANGELOG.md](https://github.com/handsontable/hyperformula/blob/master/CHANGELOG.md) file.
- [X] My changes require a documentation update.
- [ ] My changes require a migration guide.

### CODE REVIEW

**Comment by sequba** [comment]
@thilgen Thank you very much for your involvement in the project. We need to focus on the scope of the upcoming release, and currently, we have no capacity to review your pull request. I can assure you that we will check it out as soon as we can.

File: test/interpreter/function-address.spec.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      ['=ADDRESS(1,1,1, true, "")'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqualError(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+    expect(engine.getCellValue(adr('A2'))).toEqualError(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+    expect(engine.getCellValue(adr('A3'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A4'))).toEqualError(detailedError(ErrorType.VALUE, ErrorMessage.NumberCoercion))
+    expect(engine.getCellValue(adr('A5'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A6'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueLarge))
+    expect(engine.getCellValue(adr('A7'))).toEqualError(detailedError(ErrorType.NA, ErrorMessage.ArgumentMustEqual('useA1Style', 'TRUE')))
```
Comment (sequba) [inline test/interpreter/function-address.spec.ts:69]: In my opinion, we should implement the conversion to R1C1 as well. It doesn't require a lot of effort but will improve the compatibility with both Excel and Sheets.

File: test/interpreter/function-address.spec.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      ['=ADDRESS("row1","row2")'],
+      ['=ADDRESS(1,1,0)'],
+      ['=ADDRESS(1,1,5)'],
+      ['=ADDRESS(1,1,4,FALSE())'],
+      ['=ADDRESS(1,1,1, true, "")'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqualError(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+    expect(engine.getCellValue(adr('A2'))).toEqualError(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+    expect(engine.getCellValue(adr('A3'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
```
Comment (sequba) [inline test/interpreter/function-address.spec.ts:65]: Thank you very much for analyzing the differences between your implementation and Excel/Sheets. All these differences (if cannot be eliminated easily) should be added to the table in `list-of-differences.md` file.
image

File: src/interpreter/plugin/AddressPlugin.ts
Comment (sequba) [inline src/interpreter/plugin/AddressPlugin.ts:1]: This function should be documented in the `built-in-functions.md` page.

Comment (sequba) [inline src/interpreter/plugin/AddressPlugin.ts:1]: We need a changelog entry as it is a new feature.

**Review by sequba — CHANGES_REQUESTED** [review]
@thilgen thank you for another valuable contribution. It's a good implementation. It requires some polishing. Please don't feel obliged to fix all these issues if you don't want to. We can handle them on our own ;)

**Comment by thilgen** [comment]
I am happy to address these requested changes

File: src/i18n/languages/csCZ.ts [hunk 3]
```diff
@@ -19,6 +19,7 @@ const dictionary: RawTranslationPackage = {
   },
   functions: {
     FILTER: 'FILTER',
+    ADDRESS: 'ADDRESS',
```
[the remark this answers is not in the document]
Comment (thilgen) [reply src/i18n/languages/csCZ.ts]: Added translations that match those present in [Excel Functions Translator](https://support.microsoft.com/en-us/office/excel-functions-translator-f262d0c0-991c-485b-89b6-32cc8d326889).

File: docs/guide/built-in-functions.md [hunk 4]
```diff
@@ -211,19 +211,20 @@ Total number of functions: **{{ $page.functionsCount }}**
...
-| ROW         | Returns row number of a given reference or formula reference if argument not provided.                              | ROW([Reference])                                    |
-| ROWS        | Returns the number of rows in the given reference.                                                                  | ROWS(Array)                                         |
-| VLOOKUP     | Searches vertically with reference to adjacent cells to the right.                                                  | VLOOKUP(Search_Criterion, Array, Index, Sort_Order) |
+| Function ID | Description                                                                                                         | Syntax                                                                 |
+| :---------- | :------------------------------------------------------------------------------------------------------------------ | :--------------------------------------------------------------------- |
+| ADDRESS     | Returns a cell reference as a string.                                                                               | ADDRESS(Row, Column[, AbsoluteRelativeMode[, UseA1Notation[, Sheet]]]) |
```
Comment (thilgen) [inline docs/guide/built-in-functions.md:216]: This line (ADDRESS) is the only content change - Other lines are space adjustments for the wider table column

File: src/interpreter/plugin/AddressPlugin.ts
Comment (thilgen) [reply src/interpreter/plugin/AddressPlugin.ts:1]: Added `ADDRESS` to the `built-in-functions.md` page.

File: test/interpreter/function-address.spec.ts (hunk 1 above)
Comment (thilgen) [reply test/interpreter/function-address.spec.ts]: Added support for R1C1 notation - Passing TRUE() for the 4th argument will no longer result in an error. Updated unit tests; Verified coverage; Updated the information in https://github.com/handsontable/hyperformula/issues/1221.

File: test/interpreter/function-address.spec.ts (hunk 2 above)
Comment (thilgen) [reply test/interpreter/function-address.spec.ts:93]: I have added the difference in how empty strings are handled to `list-of-differences.md.`

The other category of differences - where different error types are returned when doing range checks on `FunctionPlugin` arguments - I have added to my TODO list to dig into a little deeper as I suspect that difference affects all plugins not just this one.

File: src/interpreter/plugin/AddressPlugin.ts
Comment (thilgen) [reply src/interpreter/plugin/AddressPlugin.ts:1]: added item to changelog

File: src/interpreter/plugin/AddressPlugin.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ * @license
+ * Copyright (c) 2023 Handsoncode. All rights reserved.
+ */
+
+import {columnIndexToLabel} from '../../parser/addressRepresentationConverters'
+import {ProcedureAst} from '../../parser'
+import {InterpreterState} from '../InterpreterState'
+import {InterpreterValue} from '../InterpreterValue'
+import {FunctionArgumentType, FunctionPlugin, FunctionPluginTypecheck} from './FunctionPlugin'
+import {CellError, ErrorType} from '../../Cell'
```
[the remark this answers is not in the document]
Comment (thilgen) [reply src/interpreter/plugin/AddressPlugin.ts]: fixed - waiting on rerun

File: src/interpreter/plugin/AddressPlugin.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import {ErrorMessage} from '../../error-message'
```
[the remark this answers is not in the document]
Comment (thilgen) [reply src/interpreter/plugin/AddressPlugin.ts]: fixed - waiting on rerun

File: src/interpreter/plugin/AddressPlugin.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  FullyRelative = 4,
+}
+
+export class AddressPlugin extends FunctionPlugin implements FunctionPluginTypecheck<AddressPlugin> {
+  public static implementedFunctions: ImplementedFunctions = {
+    'ADDRESS': {
+      method: 'address',
+      parameters: [
+        {argumentType: FunctionArgumentType.NUMBER, minValue: 1},
+        {argumentType: FunctionArgumentType.NUMBER, minValue: 1},
```
Comment (sequba) [inline src/interpreter/plugin/AddressPlugin.ts:25]: We should allow negative values for these arguments because they are valid in R1C1 references. For `=ADDRESS(-1, -1, 4, FALSE)` Excel returns `R[-1]C[-1]`.

Also, please add unit tests for calling `ADDRESS` with negative col and row numbers. For both R1C1 and A1 notation styles and for different absolute/relative modes.

**Review by sequba — CHANGES_REQUESTED** [review]

File: src/interpreter/plugin/AddressPlugin.ts (hunk 7 above)
Comment (thilgen) [reply src/interpreter/plugin/AddressPlugin.ts:25]: Oh - nice catch.

I had used the `FunctionPlugin` argument checking facility in this implementation because the [OpenDoc  spec pre-conditioned](http://docs.oasis-open.org/office/v1.2/os/OpenDocument-v1.2-os-part2.html#__RefHeading__1018416_715980110)  `row` and `column` arguments as being `>=1` - but you are correct to note that that Excel and Sheets allow values `<1` (**sometimes**)

**Frustratingly** - it appears that range checking used by both Excel and Sheets is variable depending on the other argument values: `row` and `column` must always be `>=1` if `A1` style is requested; but when requesting `R1C1` style the range checking depend on the requested `MODE`.

| ROW      | COL      | MODE    | STYLE     | ADDRESS   |
| -------- | ------- | --------- | --------- | ------------ | 
|-1| 	-1| 	1| 	FALSE| 	#VALUE!  |
|-1| 	-1| 	2| 	FALSE| 	#VALUE! |
| -1| 	-1| 	3| 	FALSE| 	#VALUE!  |
| -1	| -1| 	4| 	FALSE| 	R[-1]C[-1] |

| ROW      | COL      | MODE    | STYLE     | ADDRESS   |
| -------- | ------- | --------- | --------- | ------------ | 
| -1| 	1	| 1| 	FALSE| 	#VALUE! |
| -1	| 1| 	2	| FALSE| 	#VALUE! |
| -1	| 1| 	3| 	FALSE| 	R[-1]C1 |
| -1	| 1| 	4| 	FALSE| 	R[-1]C[1] |

| ROW      | COL      | MODE    | STYLE     | ADDRESS   |
| -------- | ------- | --------- | --------- | ------------ | 
| 1| 	-1| 	1| 	FALSE| 	#VALUE! |
| 1| 	-1| 	2| 	FALSE| 	R1C[-1] |
| 1| 	-1| 	3| 	FALSE| 	#VALUE! |
| 1| 	-1| 	4| 	FALSE| 	R[1]C[-1] |

So it looks like the arg checking provided by `FunctionPlugin` won't work and I'll need to roll custom argument validation based on the other argument values. 

Will work on that now - and I will add a big table test that exhaustively checks all argument combinations.

Thanks for catching this!

Comment (thilgen) [reply src/interpreter/plugin/AddressPlugin.ts]: Added support for negative (and zero) row and col indicators; Added full test matrix against Excel and Sheets

**Review by sequba — APPROVED** [review]

**Comment by sequba** [comment]
@adamu-handsontable, could you also review this PR?

File: src/interpreter/plugin/AddressPlugin.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        {argumentType: FunctionArgumentType.NUMBER, optionalArg: true, defaultValue: 1, minValue: 1, maxValue: 4},
+        {argumentType: FunctionArgumentType.BOOLEAN, optionalArg: true, defaultValue: true},
+        {argumentType: FunctionArgumentType.STRING, optionalArg: true},
+      ]
+    },
+  }
+
+  private verifyAddressArguments(row: number, col: number, abs: number, useA1Style: boolean): Maybe<CellError> {
+    if (useA1Style) {
+      if ((row < 1) || (col < 1)) {
```
Comment (wszymanski) [inline src/interpreter/plugin/AddressPlugin.ts:38]:
```suggestion
      if (row < 1 || col < 1) {
```

File: src/interpreter/plugin/AddressPlugin.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        return new CellError(ErrorType.VALUE, ErrorMessage.LessThanOne)
+      }
+    } else {
+      if (AbsStyle.FullyAbsolute == abs) {
+        if ((row < 1) || (col < 1)) {
```
Comment (wszymanski) [inline src/interpreter/plugin/AddressPlugin.ts:43]:
```suggestion
        if (row < 1 || col < 1) {
```

File: src/interpreter/plugin/AddressPlugin.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        if (row < 1) {
+          return new CellError(ErrorType.VALUE, ErrorMessage.LessThanOne)
+        }
+      } else if (AbsStyle.RowRelativeColAbsolute == abs) {
+        if (col < 1) {
+          return new CellError(ErrorType.VALUE, ErrorMessage.LessThanOne)
+        }
+      } else if (AbsStyle.FullyRelative == abs) {
+        // no range checks needed for AbsStyle.FullyRelative
+      }
```
Comment (wszymanski) [inline src/interpreter/plugin/AddressPlugin.ts:56]:
```suggestion
```

File: src/interpreter/plugin/AddressPlugin.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    }
+    return undefined
```
Comment (wszymanski) [inline src/interpreter/plugin/AddressPlugin.ts:58]:
```suggestion
```

File: src/interpreter/plugin/AddressPlugin.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  public address(ast: ProcedureAst, state: InterpreterState): InterpreterValue {
+    return this.runFunction(ast.args, state, this.metadata('ADDRESS'), (row: number, col: number, abs: number, useA1Style: boolean, sheetName: string): InterpreterValue => {
+
+      const argumentError = this.verifyAddressArguments(row, col, abs, useA1Style)
+      if (undefined !== argumentError) {
```
Comment (wszymanski) [inline src/interpreter/plugin/AddressPlugin.ts:65]:
```suggestion
      const argumentError = this.verifyAddressArguments(row, col, abs, useA1Style);

      if (argumentError !== undefined) {
```

File: src/interpreter/plugin/AddressPlugin.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        return argumentError
+      }
+
+      const colLetter = columnIndexToLabel(col-1)
+
+      let sheetPrefix = ''
+      if ((undefined !== sheetName) && (null !== sheetName)) {
```
Comment (wszymanski) [inline src/interpreter/plugin/AddressPlugin.ts:72]:
```suggestion
      const colLetter = columnIndexToLabel(col - 1);
      let sheetPrefix = '';

      if (sheetName !== undefined && sheetName !== null) {
```

File: src/interpreter/plugin/AddressPlugin.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+      if (AbsStyle.FullyRelative == abs) {
+        return useA1Style ? `${sheetPrefix}${colLetter}${row}` : `${sheetPrefix}${r1c1RowSegment}${r1c1ColSegment}`
+      } else if (AbsStyle.RowRelativeColAbsolute == abs) {
+        return useA1Style ? `${sheetPrefix}$${colLetter}${row}` : `${sheetPrefix}${r1c1RowSegment}C${col}`
+      } else if (AbsStyle.RowAbsoluteColRelative == abs) {
+        return useA1Style ? `${sheetPrefix}${colLetter}$${row}` : `${sheetPrefix}R${row}${r1c1ColSegment}`
+      } else {
+        return useA1Style ? `${sheetPrefix}$${colLetter}$${row}` : `${sheetPrefix}R${row}C${col}`
+      }
```
Comment (wszymanski) [inline src/interpreter/plugin/AddressPlugin.ts:87]:
```suggestion
      }

      return useA1Style ? `${sheetPrefix}$${colLetter}$${row}` : `${sheetPrefix}R${row}C${col}`;
```

**Review by wszymanski — COMMENTED** [review]
I don't mean to start code style wars so decide @sequba what will make code more clear from your perspective. In my opinion, some of the suggestions could be set in the linter configuration.

File: src/interpreter/plugin/AddressPlugin.ts (hunk 11 above)
Comment (sequba) [reply src/interpreter/plugin/AddressPlugin.ts:58]: I prefer the explicit return statement as this function is not `void`

**Review by wszymanski — APPROVED** [review]
Please wait until CI tests will pass.

**Comment by 171h** [comment]
I am really looking forward to this feature.🥹🥹🥹

**Comment by sequba** [comment]
> I am really looking forward to this feature.🥹🥹🥹

@171h It will be included in the upcoming release ;)
